### ISSUE TITLE
The event system needs improvement

### ISSUE THREAD

**Issue by Admicos** [Admicos/thatcord#2]
Currently, the event system is a horrible hack that tries to stuff traits into something like OOP interfaces, and use macros to wrap it all up in something sensible for the end user.

This system needs to be more idiomatic, though I haven't decided on a design yet.

Things to consider:

* We cannot stop the gateway handler loop unless we resume the gateway connection after every event, which doesn't sound efficient in the slightest.
* It should work "predictably". Meaning, the end user shouldn't need to worry about lifetimes or concurrency (except obvious stuff like using `.await`) or any other implementation detail
* Maybe more that comes to my mind later.

**Comment by Aeledfyr** [comment]
There are two strategies that I can think of for the event handling traits:  One of them is what serenity does, which is to have a handler trait with default methods
```rust
trait EventHandler {
    fn ready(&self, client: Client) {}
    fn guild_create(&self, client: Client, guild: Guild) {}
}
```
The other strategy is closer to what you did, which would be to have a trait for each event type, and then an auto implemented trait that wraps that.
```rust
#[async_trait::async_trait(?Send)]
pub trait EventHandler {
    async fn handle(
        &mut self,
        client: Client<'_>,
        event: String,
        data: serde_json::Value,
    ) -> Result<()>;
}
#[async_trait::async_trait(?Send)]
pub trait HandleReadyEvent {
    async fn handle_ready(
        &mut self,
        client: Client<'_>,
        data: ReadyEvent,
    ) -> Result<()>;
}
```
Whatever type is used, I think that a fully typed representation of the data should be passed to the user of the library, which partially limits the options.

**Comment by Admicos** [comment]
> There are two strategies that I can think of for the event handling
> traits:  One of them is what serenity does, which is to have a
> handler trait with default methods

This is probably the best approach so far, and was something I
considered, but decided against because I was dumb and forgot default
methods existed.

> Whatever type is used, I think that a fully typed representation of
> the data should be passed to the user of the library, which partially
> limits the options.

Agreed on that.

**Comment by Admicos** [comment]
> There are two strategies that I can think of for the event handling
> traits

Also, forgot to mention: Traits aren't a requirement. Something like
matching events and letting the end user handle the structure would
also be idea.

**Comment by Admicos** [comment]
On Tue, 24 Dec 2019 22:38:20 +0300
Ecmel Berk Canlıer <me@ecmelberk.com> wrote:

> Also, forgot to mention: Traits aren't a requirement. Something like
> matching events and letting the end user handle the structure would
> also be idea.

Well, traits *may* be a requirement. For the last couple hours I've
been trying to implement a `match {}` based system, and with
async/await, that's just unreasonably hard, and is equally hacky.

I'll try looking at the "single struct for all events" way, which will
probably end up as the cleanest way to do this.

**Comment by Admicos** [comment]
**Update:** I have made an initial implementation of the "single trait for all events" method on the `single-trait-events-v2` branch. The internals are still pretty hacky, but we *should* be able to re-structure those without breaking the API (though not exactly sure on that)

**Comment by Admicos** [comment]
**Update:** I figured out a better way via pattern matching which uses no internal hacks at all! It's on the `event-matching-v2` branch, and will probably be the way to go unless something happens.

**Comment by Aeledfyr** [comment]
One issue with the polling method is that it forces the events system to be single threaded;  with the traits method, we could make the methods take an immutable reference and then potentially make event handling multithreaded.  I'm not sure how much of a priority that is, but it seems like it would go well with the async/await and futures goal.
The main tradeoff is between the library controlling the event loop and the application controlling the event loop.  If the library controls it, the applications have less freedom on what they can do, but the library can make larger architectural optimizations like multithreading.  If the application controls it, the library has less control and can do fewer optimizations, but the application can control the structure.
Serenity goes fully into the first approach, which makes it hard to make scraping applications that just run once and avoid the event loop alltogether.  I haven't seen a discord library that goes fully into the other direction, but I would think that that may lead to problems further down the line relating to voice connections, and later optimizations.

**Comment by Admicos** [comment]
> One issue with the polling method is that it forces the events system to be single threaded

Yeah, especially sharding worries me with the polling approach. We _could_ have a "simple" (polling, single thread) and an "advanced" (structs, multithreaded) API for that, and people could choose for their use cases, but that might be unnecessary duplication.

>  I'm not sure how much of a priority that is, but it seems like it would go well with the async/await and futures goal.

While my own reason for making this doesn't really need multi-threaded or sharded Discord API usage, that's definitely something nice to have.
